### PR TITLE
Return the retry count, when a message fails

### PR DESCRIPTION
**Pull request by igrayson**

### CODE REVIEW

**Comment by igrayson** [comment]
This lets us more easily build logic to handle the 'first' failure differently from repeated failures.

File: src/source_sink_runner.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export interface SourceSinkRunnerConfig<T extends Message, R extends Message> {
   sink?:MessageSink<R>;
   logger?:Logger;
   delayMs?:number;
-  errorHandler?:(error:Error, log:Logger) => void;
+  errorHandler?:(error:Error, log:Logger, attemptCount:number|void) => void;
```
Comment (nevir) [inline src/source_sink_runner.ts]: You're always calling it with a `number`; doesn't look like you need `|void`

File: src/source_sink_task.ts [hunk 2]
```diff
@@ -14,7 +14,16 @@ export interface MessageTransformer<T extends Message, R extends Message> {
...
+
+  /**
+   * Called when a message could not be completed, and is expected to be retried
+   * (returned by some later getNext() call). This gives the implementation an opportunity
+   * to apply a backoff strategy to the message.
+   *
+   * If available, this returns the number of times a message has been attempted (including
+   * this attempt).
+   */
+  backoff(message:T):Promise<number|void>;
```
Comment (nevir) [inline src/source_sink_task.ts]: This always resolves to a `number` in the current implementations, from what I can tell.  Let's stick to the tightest type signatures until we absolutely have to loosen 'em, IMO.

Comment (igrayson) [inline src/source_sink_task.ts]: Nope, not all implementations resolve to a `number` -- just the ones I updated.

**Pull request merged**